### PR TITLE
Acquiring the segment reference lock on task level

### PR DESCRIPTION
**Pull request by xiaocai2333**
issue: #17523 
Signed-off-by: Cai.Zhang <cai.zhang@zilliz.com>

### CODE REVIEW

File: internal/datacoord/segment_reference_manager.go [hunk 1]
```diff
@@ -92,68 +67,55 @@ func NewSegmentReferenceManager(etcdKV kv.BaseKV, onlineIDs []UniqueID) (*Segmen
...
 
-	log.Info("Release reference lock on segments", zap.Int64s("segIDs", segIDs), zap.Int64("nodeID", nodeID))
-	locKeys := make([]string, 0)
-	for _, segID := range segIDs {
-		for _, segLock := range srm.segmentsLock[segID] {
-			if segLock.nodeID == nodeID {
-				locKeys = append(locKeys, segLock.locKey)
-			}
-		}
+	log.Info("Release reference lock by taskID", zap.Int64("taskID", taskID), zap.Int64("nodeID", nodeID))
```
[the remark this answers is not in the document]
Comment (xiaocai2333) [reply internal/datacoord/segment_reference_manager.go:105]: ok, I will pay attention to this.

File: internal/indexcoord/index_coord.go [hunk 2]
```diff
@@ -873,8 +873,10 @@ func (i *IndexCoord) watchMetaLoop() {
 	}
 }
 
-func (i *IndexCoord) tryAcquireSegmentReferLock(ctx context.Context, segIDs []UniqueID) error {
+func (i *IndexCoord) tryAcquireSegmentReferLock(ctx context.Context, buildID UniqueID, segIDs []UniqueID) error {
+	// IndexCoord use buildID instead of taskID.
 	status, err := i.dataCoordClient.AcquireSegmentLock(ctx, &datapb.AcquireSegmentLockRequest{
```
Comment (xiaofan-luan) [inline internal/indexcoord/index_coord.go:878]: question, do we ensure taskID of different?
IF not maybe nodeID + taskID can be the unique key

File: internal/datacoord/garbage_collector_test.go [hunk 3]
```diff
@@ -122,13 +122,11 @@ func Test_garbageCollector_scan(t *testing.T) {
...
 		segReferManager := &SegmentReferenceManager{
 			etcdKV: etcdKV,
-			segmentsLock: map[UniqueID][]*SegmentLock{
-				2: {
-					{
-						segmentID: 2,
-						nodeID:    1,
-						locKey:    "path",
-					},
+			segmentsLock: map[UniqueID]*datapb.SegmentReferenceLock{
```
Comment (xiaofan-luan) [inline internal/datacoord/garbage_collector_test.go:125]: I was thinking add a 
SegmentID -> task list inverted index to accelerate HasSegmentLock request

**Pull request merged**